### ISSUE TITLE
Simulation-Reconstruction chains not working for Run3 scenario using dd4hep

### ISSUE THREAD

**Issue by bsunanda** [cms-sw/cmssw#31143]
None of the steps SIM-DIGI-RECO is functional currently with dd4hep geometry definition. The area SimG4Core/Configuration/test is equipped with 7 cfg files which can help to resolve the issue:
dd4hep_ZMM_Run3_StepN_cfg.py provides the configurations for Run3 with dd4hep for SIM (N=1), DIGI (N=2), RECO (N=3)
Since none of these steps are completing satisfactorily, the corresponding cfg's with DDD are also provided as ddd_ZMM_Run3_StepN_cfg.py
The seventh cfg is for step1 of a Phase2 run scenario

Error message from Step1:
%MSG-w TkAccumulatingSensitiveDetector::createHit:  OscarMTProducer:g4SimHits  04-Aug-2020 14:45:25 CEST Run: 1 Event: 1
 theDetUnitId is not valid for TrackerHitsTEC
%MSG
----- Begin Fatal Exception 04-Aug-2020 14:45:25 CEST-----------------------
An exception of category 'TkAccumulatingSensitiveDetector::createHit' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'FEVTDEBUGoutput_step'
   [2] Prefetching for module PoolOutputModule/'FEVTDEBUGoutput'
   [3] Calling method for module OscarMTProducer/'g4SimHits'
Exception Message:
cannot get theDetUnitId for G4Track 5262
----- End Fatal Exception -------------------------------------------------

Error message from Step2:
----- Begin Fatal Exception 04-Aug-2020 15:03:30 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Running EventSetup component DTGeometryESModule/'
Exception Message:
No data of type "DDCompactView" with label "" in record "IdealGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------

Error message from Step3:
cmsRun: /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/1d6ef0028b473072ff15e3d6bc5122f8/opt/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-03-2300/src/Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc:67: TrackerGeometry::TrackerGeometry(const GeometricDet*): Assertion `subdetgd[i]->geographicalId().subdetId() > 0 && subdetgd[i]->geographicalId().subdetId() < 7' failed.

Please note: 
(1) The job fails on encountering the first error. Till these are resolved we may not find other hidden errors
(2) Inputs for testing step2 or step3 are produced using the DDD examples

**Comment by bsunanda** [comment]
@vargasa @slomeo @cvuosalo @ianna @kpedro88 @fabiocos @civanch @perrotta @watson-ij Please watch and take action

**Comment by makortel** [comment]
assign geometry

**Comment by slomeo** [comment]
@bsunanda : If I understand well, this PR is what I have to do, by using  @fabiocos ' s PR #30003 , when all DD4hep migrations will be merged (till now my PR #31043 ,  that's the last PR for DD4HEP migration for MUON, is under review waiting that the new Geometry repo will be created), Am I correct? Please @fabiocos can you comment?

**Comment by bsunanda** [comment]
@slomeo I believe the tracker has to address that first. They have issues for GEN-SIM as well as RECO step. Muon gives rise to first failure in the DIGI step. You can take a look at SimG4Core/Configuration/test area. Please generate step1.root using the GEN-SIM (step1) cfg file for DDD and then try DIGI (step2) cfg with dd4hep. You need not try the full workflow as made by Fabio right now.

**Comment by slomeo** [comment]
@bsunanda : tks, I'll wait for the Tracker update.

**Comment by vargasa** [comment]
Hi all, from a previous [private discussion](https://drive.google.com/file/d/1lY_N9Y22zxZUqj1d-8TtWE7N3xEyqLG1/view?usp=sharing):

@ghugo83 pointed out that these errors may be seen because of overlaps still present in CMS geometry (not necessarily in the tracker, the remaining overlap issues with the tracker was solved back in April).

From the same discussion @bsunanda commented:

> A large fraction of those 1330 overlaps - 990 are due to overlap of the volume elements (SFBX/SFBY) of preshower detector extruded from teir mother volume. These elements are mother of the sensitive detectors (SFSX/SFSY).

From @rchatter:
> The last time we checked this our understanding was that there was no overlap in any of the active elements. Then related to the PreShower in the context of the Boolean solids we agreed not to re-look at the code largely due to the fact that this subdetector would be replaced for HL-LHC.

@bsunanda , @rchatter, @cvuosalo, @civanch, from that discussion it is not clear to me if those issues were addressed. Can you please briefly comment/confirm the present issue is not related with that anymore?

**Comment by civanch** [comment]
@vargasa , when we discussed these pre-shower problems it was said, that the code for ECAL it too big and difficult to debug. 

Ianna mentioned that after migration to DD4Hep it would be possible make a fix.

**Comment by mmusich** [comment]
@vargasa 
out of curiosity I had a look to the "Tracker" issue with step3.
With these changes:

```diff
diff --git a/Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc b/Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc
index 1c6cc137d00..024ae3c4834 100644
--- a/Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc
+++ b/Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc
@@ -62,6 +62,8 @@ TrackerGeometry::TrackerGeometry(GeometricDet const* gd) : theTrackerDet(gd) {
   }
   GeometricDet::ConstGeometricDetContainer subdetgd = gd->components();
 
+  std::cout<<"size: " << subdetgd.size() << std::endl;
+
   LogDebug("BuildingSubDetTypeMap") << "GeometriDet and GeomDetEnumerators enumerator values of the subdetectors";
   for (unsigned int i = 0; i < subdetgd.size(); ++i) {
     assert(subdetgd[i]->geographicalId().subdetId() > 0 && subdetgd[i]->geographicalId().subdetId() < 7);
diff --git a/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerBuilder.cc b/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerBuilder.cc
index ccc2f248b23..13fabdced21 100644
--- a/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerBuilder.cc
+++ b/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerBuilder.cc
@@ -16,6 +16,8 @@ void CmsTrackerBuilder<FilteredView>::buildComponent(FilteredView& fv, Geometric
   CmsTrackerPixelPhase1EndcapBuilder<FilteredView> theCmsTrackerPixelPhase1EndcapBuilder;
   CmsTrackerPixelPhase2EndcapBuilder<FilteredView> theCmsTrackerPixelPhase2EndcapBuilder;
 
+  std::cout<<"CmsTrackerBuilder::buildComponent() " <<  s << std::endl;
+
   GeometricDet* subdet = new GeometricDet(&fv,
                                           CmsTrackerLevelBuilder<FilteredView>::theCmsTrackerStringToEnum.type(
                                               ExtractStringFromDDD<FilteredView>::getString(s, &fv)));
diff --git a/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerLevelBuilder.cc b/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerLevelBuilder.cc
index e98e35b1f24..451d54ad612 100644
--- a/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerLevelBuilder.cc
+++ b/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerLevelBuilder.cc
@@ -141,7 +141,7 @@ template <class FilteredView>
 void CmsTrackerLevelBuilder<FilteredView>::build(FilteredView& fv,
                                                  GeometricDet* tracker,
                                                  const std::string& attribute) {
-  LogTrace("GeometricDetBuilding")  //<< std::string(3 * fv.history().size(), '-') << "+ "
+  edm::LogPrint("GeometricDetBuilding")  //<< std::string(3 * fv.history().size(), '-') << "+ "
       << ExtractStringFromDDD<FilteredView>::getString(attribute, &fv) << " " << tracker->type() << " "
       << tracker->name() << std::endl;
 
@@ -150,6 +150,7 @@ void CmsTrackerLevelBuilder<FilteredView>::build(FilteredView& fv,
   while (doLayers) {
     buildComponent(fv, tracker, attribute);
     doLayers = fv.nextSibling();  // go to next layer
+    std::cout << "doLayers: " << doLayers << std::endl;
   }
 
   fv.parent();
```

running `/SimG4Core/Configuration/test/dd4hep_ZMM_Run3_Step3_cfg.py` I get:

```bash
FullTracker 0 Tracker
CmsTrackerBuilder::buildComponent() TkDDDStructure
PixelPhase1Barrel 101 PixelBarrel
PixelBarrelLayer 8 pixbarlayer0:PixelBarrelLayer0
PixelBarrelLadder 14 PixelBarrelLadderFull0
doLayers: 0
doLayers: 0
doLayers: 0
doLayers: 0
size: 1
cmsRun: /tmp/musich/CMSSW_11_2_X_2020-08-17-1100/src/Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc:69: TrackerGeometry::TrackerGeometry(const GeometricDet*): Assertion `subdetgd[i]->geographicalId().subdetId() > 0 && subdetgd[i]->geographicalId().subdetId() < 7' failed.
```
while in the ddd case: `/SimG4Core/Configuration/test/ddd_ZMM_Run3_Step3_cfg.py`
I get:
```bash
FullTracker 0 Tracker
CmsTrackerBuilder::buildComponent() TkDDDStructure
PixelPhase1Barrel 101 PixelBarrel
PixelBarrelLayer 8 PixelBarrelLayer0
PixelBarrelLadder 14 PixelBarrelLadderFull0
doLayers: 1
doLayers: 1
doLayers: 1
doLayers: 1
doLayers: 1
doLayers: 1
doLayers: 1
doLayers: 0
doLayers: 1
PixelBarrelLadder 14 PixelBarrelLadderFull0
doLayers: 1
.
.
...
```

looks like something funny is going on with `FilteredView::nextSibling()` in the dd4hep case.

**Comment by vargasa** [comment]
Hi @mmusich thanks, do you still have your workspace? can you try to run it adding:
```
process.MessageLogger.categories.append("DDFilteredView")
```

**Comment by bsunanda** [comment]
@vargasa @mmusich The looping over fv's are totally different between DDD and DD4Hep. The best way to loop over fv's in dd4hep is to use the loop 
while (fv.firstChild()) {
}
and not to use any other operation on fv by going up/down/next. I had to take care of this for some muon detectors. If you want I can take a look into this.

**Comment by mmusich** [comment]
> do you still have your workspace? can you try to run it adding:

@vargasa yes I did, but nothing significant pops out, full log at https://paste.in/u8qTI9

**Comment by mmusich** [comment]
@bsunanda 

>  and not to use any other operation on fv by going up/down/next. I had to take care of this for some muon detectors. If you want I can take a look into this.

yes, thanks!

**Comment by vargasa** [comment]
Hi all, some previously hidden overlaps were found in the outer barrel (Tracker), just made a PR https://github.com/cms-sw/cmssw/pull/31181 for that . I'll also open an issue with a list of the overlaps that are currently found so we can disentangle the source of the current issue

**Comment by vargasa** [comment]
> @vargasa yes I did, but nothing significant pops out, full log at https://paste.in/u8qTI9

`Search for first child failed` tell us that probably an important piece of code is being ignored :sweat_smile:

**Comment by bsunanda** [comment]
@ghugo83 Please have a look into this issue

**Comment by ghugo83** [comment]
@bsunanda 
Thanks for forwarding this, will have a look.

**Comment by ghugo83** [comment]
Starting to have a look now.
Hmm @vargasa the overlaps I mentioned were for another issue, for a random crash which was obtained on a workflow based on old DD.

First just wondering: when you migrate to DD4hep, did you print all Tracker DetIds and coordinates, to make sure they are identical than with old DD?
That is the first thing to do, for all volumes and all scenarios. Otherwise you cannot expect the steps downstream to work smoothly ;p
Did you do such tests?
I did these types of tests when migrating PPS for example, and there were defo a lot of points to address before getting a 1:1 comparison.

**Comment by vargasa** [comment]
Hi @ghugo83 please refer to: https://hypernews.cern.ch/HyperNews/CMS/get/geometry/359/3/1/1/1/2.html

**Comment by ghugo83** [comment]
Did some investigations.
First thing is that, presently, the Tracker geometry built with DD4hep is a... 1-sensor Tracker.

Basically, the first sensor (in TrackerNumberingBuilder order) is built (a pixel sensor), then the navigation crashes.
This explains the error messages where detUnitId are not found, subdetId not found, etc: there is nearly no sensor.

Am first fixing navigation, but I presume from this that there will be lots of points to look at, as there is actually no real 'DD4hep' Tracker geometry yet. 

Going on the fixes on the fly, will come back when I have advanced a big block.

**Comment by ghugo83** [comment]
> looks like something funny is going on with `FilteredView::nextSibling()` in the dd4hep case.

@mmusich +1 Yes exactly.
The geo is built with one sensor, then the navigation crashes, leaving the Tracker geo with one sensor in total.
I presume that point had not been fixed then.
Anyway I am developing in a private branch where am fixing all this now.

**Comment by ghugo83** [comment]
@vargasa Thanks can you send any other PR (pending / done) on this topic ? Cheers

**Comment by ghugo83** [comment]
> Hi @ghugo83 please refer to: https://hypernews.cern.ch/HyperNews/CMS/get/geometry/359/3/1/1/1/2.html

@vargasa Thanks, but can we keep this discussion here, in this github issue ?
HN is really spamming a lot of people, while a github issue is more appropriate I think :)

Plus email groups are so much easier to send emails to / rather than having to connect to HN.

**Comment by vargasa** [comment]
> Could you please send any other PR (done / pending) on this topic, if any ?

No pending PRs from this side.

Regarding the Navigation, I just noticed an additional detail after #31337 : 

```
dd4hep:
* Doesn't go through the blades as dd does
/OCMS_1/CMSE_1/Tracker_1/PixelForwardZplus_1
/OCMS_1/CMSE_1/Tracker_1/PixelForwardZplus_1/pixfwdDisks:PixelForwardDiskZplus_1
/OCMS_1/CMSE_1/Tracker_1/PixelForwardZplus_1/pixfwdDisks:PixelForwardDiskZplus_2
/OCMS_1/CMSE_1/Tracker_1/PixelForwardZplus_1/pixfwdDisks:PixelForwardDiskZplus_3
/OCMS_1/CMSE_1/Tracker_1/PixelForwardZminus_1
/OCMS_1/CMSE_1/Tracker_1/PixelForwardZminus_1/pixfwdDisks:PixelForwardDiskZminus_1
/OCMS_1/CMSE_1/Tracker_1/PixelForwardZminus_1/pixfwdDisks:PixelForwardDiskZminus_2
/OCMS_1/CMSE_1/Tracker_1/PixelForwardZminus_1/pixfwdDisks:PixelForwardDiskZminus_3

dd:
/cms:OCMS[0]/cms:CMSE[1]/tracker:Tracker[1]/pixfwd:PixelForwardZplus[1]/pixfwdDisks:PixelForwardDiskZplus[1]/pixfwdbladeInnerZplus:PixelForwardBlade[0]/pixfwdbladeInnerZplus:PixelForwardModule[1]/pixfwdbladeInnerZplus:PixelForwardWafer[1]
PixelEndcapPanel 18 PixelForwardWafer
/cms:OCMS[0]/cms:CMSE[1]/tracker:Tracker[1]/pixfwd:PixelForwardZplus[1]/pixfwdDisks:PixelForwardDiskZplus[1]/pixfwdbladeInnerZplus:PixelForwardBlade[0]/pixfwdbladeInnerZplus:PixelForwardModule[1]/pixfwdbladeInnerZplus:PixelForwardWafer[1]/pixfwdbladeInnerZplus:PixelForwardSensor[1]
/cms:OCMS[0]/cms:CMSE[1]/tracker:Tracker[1]/pixfwd:PixelForwardZplus[1]/pixfwdDisks:PixelForwardDiskZplus[1]/pixfwdbladeInnerZplus:PixelForwardBlade[0]/pixfwdbladeInnerZplus:PixelForwardModule[1]/pixfwdbladeInnerZplus:PixelForwardWafer[1]
CmsTrackerLevelBuilder::build  Building: /cms:OCMS[0]/cms:CMSE[1]/tracker:Tracker[1]/pixfwd:PixelForwardZplus[1]/pixfwdDisks:PixelForwardDiskZplus[1]/pixfwdbladeInnerZplus:PixelForwardBlade[0]/pixfwdbladeInnerZplus:PixelForwardModule[2]/pixfwdbladeInnerZplus:PixelForwardWafer[1]

dd4hep:
Goes into: TIBModule (dd doesn't) same with TIB, TIBLayer1, TIDF, TIDB
/OCMS_1/CMSE_1/Tracker_1/TIB_1/tiblayer0:TIBLayer0_1/tiblayer0:TIBLayer0Down_1/TIBString0Lo1_1/TIBString0LoMin1_1/TIBModule0B_1/TIBWaferSter0_1
/OCMS_1/CMSE_1/Tracker_1/TIB_1/tiblayer0:TIBLayer0_1/tiblayer0:TIBLayer0Down_1/TIBString0Lo1_1/TIBString0LoMin1_1/TIBModule0B_1/TIBWaferSter0_1/TIBActiveSter0_1
/OCMS_1/CMSE_1/Tracker_1/TIB_1/tiblayer0:TIBLayer0_1/tiblayer0:TIBLayer0Down_1/TIBString0Lo1_1/TIBString0LoMin1_1/TIBModule0B_2/TIBWaferSter0_1
/OCMS_1/CMSE_1/Tracker_1/TIB_1/tiblayer0:TIBLayer0_1/tiblayer0:TIBLayer0Down_1/TIBString0Lo1_1/TIBString0LoMin1_1/TIBModule0B_2/TIBWaferSter0_1/TIBActiveSter0_1
...

dd:
/cms:OCMS[0]/cms:CMSE[1]/tracker:Tracker[1]/tib:TIB[1]/tiblayer0:TIBLayer0[1]/tiblayer0:TIBLayer0Down[1]/tibstring0:TIBString0Lo1[1]/tibstring0ll:TIBString0LoMin1[1]/tibmodule0b:TIBModule0B[1]
/cms:OCMS[0]/cms:CMSE[1]/tracker:Tracker[1]/tib:TIB[1]/tiblayer0:TIBLayer0[1]/tiblayer0:TIBLayer0Down[1]/tibstring0:TIBString0Lo1[1]/tibstring0ll:TIBString0LoMin1[1]/tibmodule0b:TIBModule0B[2]
...
```

Full Logs:
DD: https://drive.google.com/file/d/1LBDv_AnzbYTFkT8ci8Vg92uO3sIQKCsw/view
DD4hep: https://drive.google.com/file/d/1Dpq3VlUGm6gYV7NuymZb4vcRoJpxU0zh/view

**Comment by ghugo83** [comment]
Finished with PPS ;ppppp With the XMLs PR this morning.

So finally looking properly at this.

In old DD, the navigation in the Tracker is like this: https://github.com/cms-sw/cmssw/blob/CMSSW_11_1_X/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerLevelBuilder.cc#L148  

```
template <class FilteredView>
void CmsTrackerLevelBuilder<FilteredView>::build(FilteredView& fv,
                                                 GeometricDet* tracker,
                                                 const std::string& attribute) {
  LogTrace("GeometricDetBuilding")  //<< std::string(3 * fv.history().size(), '-') << "+ "
      << ExtractStringFromDDD<FilteredView>::getString(attribute, &fv) << " " << tracker->type() << " "
      << tracker->name() << std::endl;

  bool doLayers = fv.firstChild();  // descend to the first Layer

  while (doLayers) {
    buildComponent(fv, tracker, attribute);
    doLayers = fv.nextSibling();  // go to next layer
  }

  fv.parent();

  sortNS(fv, tracker);
}
```

If you look at the PPS PR, it is exactly the same as the old DD navigation in PPS: https://github.com/cms-sw/cmssw/pull/31383/files#diff-1565827770ef64555939430a61bf05e8R29
```
void detgeomdescbuilder::buildDetGeomDescDescendants(DDFilteredView& fv, DetGeomDesc* geoInfo) {
  // Leaf
  if (!fv.firstChild())
    return;

  do {
    // Create node, and add it to the geoInfoParent's list.
    DetGeomDesc* child = new DetGeomDesc(fv);
    geoInfo->addComponent(child);

    // Recursion
    buildDetGeomDescDescendants(fv, child);
  } while (fv.nextSibling());

  fv.parent();
}
```

For PPS, this was solved like this for DD4hep navigation:
https://github.com/cms-sw/cmssw/pull/31383/files#diff-1565827770ef64555939430a61bf05e8R49 
```
std::unique_ptr<DetGeomDesc> detgeomdescbuilder::buildDetGeomDescFromCompactView(
  
  if (fv.next(0) == false) {
    edm::LogError("PPSGeometryESProducer") << "Filtered view is empty. Cannot build.";
  }

  // Geo info: root node.
  auto geoInfoRoot = std::make_unique<DetGeomDesc>(fv);

  // Construct the tree of children geo info (DetGeomDesc).
  do {
    // Create node, and add it to the geoInfoRoot's list.
    DetGeomDesc* child = new DetGeomDesc(fv);
    geoInfoRoot->addComponent(child);
  } while (fv.next(0));

  return geoInfoRoot;
}
```

Doing this on the Tracker level builder, it works.
One important difference is that for the Tracker, it is recursive, relying on the abstract function buildComponent, which is virtual pure. 
So you cannot easily add the 'come back up' point (unless you add it to all derived classes derivedClass:: buildComponent, which is obviously ugly).
So the modification of @ianna in FilteredView makes sense, to be able to handle the edge case.

**Comment by cvuosalo** [comment]
+1

**Comment by cvuosalo** [comment]
@bsunanda I am running DD4hep workflows successfully to completion, so I think this issue is fixed. Could you please close it?

**Comment by bsunanda** [comment]
Simulation-Reconstruction chain works for Run3 scenario. Have to investigate Phase2 from now.
